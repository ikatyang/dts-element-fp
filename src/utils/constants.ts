export const selectable_default = false;
export const placeholder_default = false;

export const get_interface_name_default = (index: number) =>
  `CurriedFunction${index + 1}`;

export const get_interface_generic_name_default = (index: number) =>
  `T${index + 1}`;
export const get_interface_generic_return_name_default = () => 'R';

export const get_interface_parameter_name_default = (index: number) =>
  `v${index + 1}`;
export const get_interface_parameter_placeholder_name_default = (
  index: number,
) => `_${index + 1}`;

export const get_placeholder_type_default = () => 'PH';

export const get_function_parameter_placeholder_name_default = (name: string) =>
  `_${name}`;
export const get_function_selectable_selector_name_default = () => '$SEL';
export const get_function_selectable_kind_name_default = () => '$KIND';
export const get_function_type_name_default = (name: string, mask: string) =>
  `${name}_${mask}`;
