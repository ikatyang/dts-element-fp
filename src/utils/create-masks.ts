import * as R from 'ramda';
import { left_pad } from './left-pad';

/**
 * - `masks[0]` -> `"000"`
 * - `masks[1]` -> `"100"`
 * - `masks[2]` -> `"010"`
 * - ...
 */
export const create_masks = (length: number) =>
  R.repeat(0, 2 ** length).map((_, index) =>
    left_pad(index.toString(2), '0', length)
      .split('')
      .reverse()
      .join(''),
  );
