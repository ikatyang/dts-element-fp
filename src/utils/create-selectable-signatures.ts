import * as dts from 'dts-element';
import {
  get_function_selectable_kind_name_default,
  get_function_selectable_selector_name_default,
} from './constants';

export interface CreateSelectableSignaturesOptions {
  get_selectable_kind_name?: () => string;
  get_selectable_selector_name?: () => string;
}

export interface CreateSelectableSignaturesInternalOptions
  extends CreateSelectableSignaturesOptions {
  /**
   * - normal: `undefined`
   * - various: `string`
   */
  group?: string;
  is_placeholder: (value: dts.IType) => boolean;
}

export function create_selectable_signatures(
  signatures: dts.IObjectMember[],
  options: CreateSelectableSignaturesInternalOptions,
): dts.IObjectMember[] {
  // istanbul ignore next
  const {
    group,
    is_placeholder,
    get_selectable_kind_name = get_function_selectable_kind_name_default,
    get_selectable_selector_name = get_function_selectable_selector_name_default,
  } = options;

  return signatures.map(signature => {
    const function_type = (signature.owned as dts.IFunctionDeclaration).type!;

    // istanbul ignore next
    const { parameters = [] } = function_type;

    const selectable_type = parameters
      .map(parameter => (is_placeholder(parameter.type!) ? '0' : '1'))
      .join('');

    return dts.create_object_member({
      owned: dts.create_function_declaration({
        name: undefined,
        type: dts.create_function_type({
          parameters: [],
          generics: [
            dts.create_generic_declaration({
              name: get_selectable_selector_name(),
              extends: dts.create_literal_type({ value: selectable_type }),
            }),
            ...(group === undefined
              ? []
              : [
                  dts.create_generic_declaration({
                    name: get_selectable_kind_name(),
                    extends: dts.create_literal_type({ value: group }),
                  }),
                ]),
          ],
          return: function_type,
        }),
      }),
    });
  });
}
