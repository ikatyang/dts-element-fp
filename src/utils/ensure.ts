export function ensure<T>(value: any, is: (x: any) => x is T): T {
  const is_something = is(value);
  // istanbul ignore next
  if (!is_something) {
    throw new Error('Invalid value');
  }
  return value;
}
