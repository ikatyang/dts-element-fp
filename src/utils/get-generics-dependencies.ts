import * as dts from 'dts-element';
import { has } from './has';

import { IGenericDeclaration } from 'dts-element';

export const get_generics_dependencies = (
  generics: dts.IGenericDeclaration[],
) =>
  generics.map(generic => [
    generic,
    ...generics.filter(current_generic =>
      has(generic, {
        kind: dts.ElementKind.GeneralType,
        name: current_generic.name,
      }),
    ),
  ]);
