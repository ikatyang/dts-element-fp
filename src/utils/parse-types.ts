import * as dts from 'dts-element';

export function parse_types(
  types_string: string,
): Record<string, dts.IFunctionType> {
  const function_declarations = dts.parse(types_string)
    .members as dts.IFunctionDeclaration[];
  return function_declarations.reduce<Record<string, dts.IFunctionType>>(
    (current, function_declaration) => ({
      ...current,
      [function_declaration.name!]: function_declaration.type!,
    }),
    {},
  );
}
