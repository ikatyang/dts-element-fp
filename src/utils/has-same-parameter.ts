import * as dts from 'dts-element';
import * as R from 'ramda';
import { get_generics_dependencies } from './get-generics-dependencies';
import { has } from './has';

// tslint:disable-next-line:no-unused-variable no-duplicate-imports
import { IFunctionType } from 'dts-element';

export const has_same_parameter = (
  function_type_1: dts.IFunctionType,
  function_type_2: dts.IFunctionType,
  parameter_index: number,
) => {
  const [first, second] = [function_type_1, function_type_2].map(
    function_type => {
      const { parameters = [], generics = [] } = function_type;
      const parameter_type = parameters[parameter_index].type!;
      return { generics, parameter_type };
    },
  );

  if (!R.equals(first.parameter_type, second.parameter_type)) {
    // suppose type that looks different should have different meaning
    return false;
  }

  // check if the same generic name has the same meaning

  const [generic_tree_1, generic_tree_2] = [first, second].map(data => {
    const { generics, parameter_type } = data;
    const parameter_generics_and_indexes = generics
      .map((generic, index) => ({ generic, index }))
      .filter(({ generic }) =>
        has(parameter_type, {
          kind: dts.ElementKind.GeneralType,
          name: generic.name,
        }),
      );
    const generics_dependencies = get_generics_dependencies(generics);
    return parameter_generics_and_indexes.reduce<
      Record<string, dts.IGenericDeclaration[]>
    >(
      (current, { generic, index }) => ({
        ...current,
        [generic.name]: generics_dependencies[index]
          .slice()
          .sort((a, b) => a.name.localeCompare(b.name)),
      }),
      {},
    );
  });

  return R.equals(generic_tree_1, generic_tree_2);
};
