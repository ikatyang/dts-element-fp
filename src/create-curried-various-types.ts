import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  create_curried_types,
  CreateCurriedDeclarationsOptions,
} from './create-curried-types';
import {
  get_placeholder_type_default,
  selectable_default,
} from './utils/constants';
import { create_masks } from './utils/create-masks';
import { create_selectable_signatures } from './utils/create-selectable-signatures';
import { has_same_parameter } from './utils/has-same-parameter';

/**
 * @hidden
 */
export function create_various_curried_types(
  name: string,
  types: { [kind: string]: dts.IFunctionType },
  options: CreateCurriedDeclarationsOptions = {},
): dts.ITypeDeclaration[] {
  const {
    selectable = selectable_default,
    get_placeholder_type = get_placeholder_type_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const keys = Object.keys(types);

  const parameters_length = keys.reduce<number>(
    (length, key) =>
      length === -1 || length === types[key].parameters!.length
        ? types[key].parameters!.length
        : (() => {
            throw new Error(`curried function types must be the same length`);
          })(),
    -1,
  );

  const curried_types_declarations = keys.reduce<dts.ITypeDeclaration[][]>(
    (reduced, key) => [
      ...reduced,
      create_curried_types(`${name}_${key}`, types[key], {
        ...options,
        selectable: false,
      }),
    ],
    [],
  );

  const masks = create_masks(parameters_length);

  const function_types = keys.reduce<dts.IFunctionType[]>(
    (reduced, key) => [...reduced, types[key]],
    [],
  );

  const non_conflicts = R.repeat(0, parameters_length).map((_, index) => {
    const main_function_type = function_types[0];
    return function_types
      .slice(1)
      .every(current_function_type =>
        has_same_parameter(main_function_type, current_function_type, index),
      );
  });

  const conflict_declarations: dts.ITypeDeclaration[][] = [];
  const conflict_declarations_selectables: dts.IObjectMember[][][] = [];

  const non_conflict_rebases: { [name: string]: string } = {};
  const non_conflict_declarations: dts.ITypeDeclaration[] = [];
  const non_conflict_declarations_selectables: dts.IObjectMember[][] = [];

  curried_types_declarations[0].forEach((curried_type_declaration, index) => {
    const is_return_signature =
      index === curried_types_declarations[0].length - 1;
    const mask = masks[index];
    if (
      !is_return_signature &&
      mask
        .split('')
        .every((bit, bit_index) => bit === '0' || non_conflicts[bit_index])
    ) {
      const rebase_name = `${name}_${mask}`;
      non_conflict_declarations.push(
        dts.create_type_declaration({
          name: rebase_name,
          generics: curried_type_declaration.generics,
          type: dts.create_object_type({
            members: curried_types_declarations
              .map(
                curried_types_declaration =>
                  (curried_types_declaration[index].type! as dts.IObjectType)
                    .members! as dts.IObjectMember[],
              )
              .reduce((a, b) => [...a, ...b], []),
          }),
        }),
      );
      non_conflict_declarations_selectables.push(
        curried_types_declarations
          .map((curried_types_declaration, type_index) =>
            create_selectable_signatures(
              (curried_types_declaration[index].type! as dts.IObjectType)
                .members as dts.IObjectMember[],
              {
                group: keys[type_index],
                is_placeholder,
                get_selectable_kind_name,
                get_selectable_selector_name,
              },
            ),
          )
          .reduce((a, b) => [...a, ...b], []),
      );
      curried_types_declarations
        .map(x => x[index])
        .forEach(type_declaration => {
          non_conflict_rebases[type_declaration.name] = rebase_name;
        });
    } else {
      conflict_declarations.push(
        curried_types_declarations.map(
          curried_type_declarations => curried_type_declarations[index],
        ),
      );
      conflict_declarations_selectables.push(
        curried_types_declarations.map(curried_type_declarations => {
          const members = (curried_type_declarations[index]
            .type as dts.IObjectType).members;
          return members !== undefined && members.every(dts.is_object_member)
            ? create_selectable_signatures(members as dts.IObjectMember[], {
                get_selectable_kind_name,
                get_selectable_selector_name,
                is_placeholder,
              })
            : [];
        }),
      );
    }
  });

  const mixed_type_declarations_selectables = [
    ...non_conflict_declarations_selectables,
    ...conflict_declarations_selectables.reduce((a, b) => [...a, ...b], []),
  ];

  return [
    ...non_conflict_declarations,
    ...conflict_declarations.reduce((a, b) => [...a, ...b], []),
  ].map((type_declaration, index) => {
    const object_type = type_declaration.type;
    if (
      !dts.is_object_type(object_type) ||
      object_type.members!.some(member => !dts.is_object_member(member))
    ) {
      return type_declaration;
    }
    const repeated_function_types: { [name: string]: dts.IFunctionType } = {};
    const members: dts.IObjectMember[] = [];
    object_type.members!.forEach(
      (member: dts.IObjectMember, member_index, origin_members) => {
        const function_type = (member.owned as dts.IFunctionDeclaration).type!;
        const function_return_type = function_type.return! as dts.IGeneralType;
        if (function_return_type.name in non_conflict_rebases) {
          repeated_function_types[
            non_conflict_rebases[function_return_type.name]
          ] = function_type;
        } else {
          members.push(
            member,
            ...(selectable && origin_members.length > 1
              ? [mixed_type_declarations_selectables[index][member_index]]
              : []),
          );
        }
      },
    );
    const non_conflict_merged_members = Object.keys(
      repeated_function_types,
    ).map(key => {
      const function_type = repeated_function_types[key];
      const function_return_type = function_type.return as dts.IGeneralType;
      return dts.create_object_member({
        owned: dts.create_function_declaration({
          name: undefined,
          type: {
            ...function_type,
            return: dts.create_general_type({
              name: key,
              generics: function_return_type.generics,
            }),
          },
        }),
      });
    });

    members.push(...non_conflict_merged_members);
    if (selectable) {
      members.push(
        ...create_selectable_signatures(non_conflict_merged_members, {
          is_placeholder,
          get_selectable_kind_name,
          get_selectable_selector_name,
        }),
      );
    }
    object_type.members = members;

    return type_declaration;
  });
}
