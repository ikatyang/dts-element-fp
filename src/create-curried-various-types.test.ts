import * as dts from 'dts-element';
import { create_various_curried_types } from './create-curried-various-types';
import { parse_types } from './utils/parse-types';

const test_cases: Record<string, string> = {
  'non-generics': `
    function number(a: number, b: number): boolean;
    function string(a: string, b: string): boolean;
  `,
  'generics + multi-parameters': `
    function string(str: string): string;
    function list<T>(list: List<T>): T | undefined;
  `,
  'generics + multi-parameters + return-generics': `
    function list<T, U>(fn: Morphism<T, U>, list: List<T>): U[];
    function functor<T, U>(fn: Morphism<T, U>, functor: Functor<T>): Functor<U>;
  `,
};

it('should throw error if those function types have different length', () => {
  expect(() =>
    create_various_curried_types(
      '$',
      parse_types(`
    function $1(a: any): any;
    function $2(a: any, b: any): any;
  `),
    ),
  ).toThrowError();
});

describe('merge-non-conflicts', () => {
  it('should return correctly with mergable types', () => {
    expect(
      dts.emit(
        create_various_curried_types(
          '$',
          parse_types(`
            function list<T, U extends List<T>>(fn: (x: T) => T, container: U): U;
            function object<T, U extends Dictionary<T>>(fn: (x: T) => T, container: U): U;
          `),
        ),
      ),
    ).toMatchSnapshot();
  });
  it('should return correctly with un-mergable types', () => {
    expect(
      dts.emit(
        create_various_curried_types(
          '$',
          parse_types(`
            function list<T, U extends List<T>>(fn: (x: T) => T, container: U): U;
            function object<X, Y extends Dictionary<X>>(fn: (x: X) => X, container: Y): Y;
          `),
        ),
      ),
    ).toMatchSnapshot();
  });
});

Object.keys(test_cases).forEach(case_name => {
  describe(case_name, () => {
    it('should transform correctly without any options', () => {
      expect(emit_curried_types()).toMatchSnapshot();
    });

    it('should transform correctly with selectable', () => {
      expect(
        emit_curried_types({
          selectable: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform correctly with placeholder', () => {
      expect(
        emit_curried_types({
          placeholder: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform correctly with every option', () => {
      expect(
        emit_curried_types({
          selectable: true,
          placeholder: true,
        }),
      ).toMatchSnapshot();
    });
  });

  function emit_curried_types(
    options: { selectable?: boolean; placeholder?: boolean } = {},
  ) {
    const test_case = test_cases[case_name];
    const function_types = parse_types(test_case);

    const curried_types = create_various_curried_types(
      '$',
      function_types,
      options,
    );
    const top_level_element = dts.create_top_level_element({
      members: curried_types,
    });
    return dts.emit(top_level_element);
  }
});
