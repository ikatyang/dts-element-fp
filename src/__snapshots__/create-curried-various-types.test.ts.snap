// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generics + multi-parameters + return-generics should transform correctly with every option 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"list\\">(): <T>(_fn: PH, list: List<T>) => $_list_01<T>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <$SEL extends \\"01\\", $KIND extends \\"functor\\">(): <T>(_fn: PH, functor: Functor<T>) => $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with placeholder 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <T>(_fn: PH, list: List<T>): $_list_01<T>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <T>(_fn: PH, functor: Functor<T>): $_functor_01<T>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with selectable 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"list\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => $_list_11<U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"11\\", $KIND extends \\"functor\\">(): <T, U>(fn: Morphism<T, U>, functor: Functor<T>) => $_functor_11<U>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): (list: List<T>) => $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
    <$SEL extends \\"1\\", $KIND extends \\"functor\\">(): (functor: Functor<T>) => $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly without any options 1`] = `
"type $_00 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): $_list_11<U>;
    <T, U>(fn: Morphism<T, U>, functor: Functor<T>): $_functor_11<U>;
    <T, U>(fn: Morphism<T, U>): $_10<T, U>;
};
type $_10<T, U> = {
    (list: List<T>): $_list_11<U>;
    (functor: Functor<T>): $_functor_11<U>;
};
type $_list_01<T> = {
    <U>(fn: Morphism<T, U>): $_list_11<U>;
};
type $_functor_01<T> = {
    <U>(fn: Morphism<T, U>): $_functor_11<U>;
};
type $_list_11<U> = U[];
type $_functor_11<U> = Functor<U>;
"
`;

exports[`generics + multi-parameters should transform correctly with every option 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => $_string_1;
    <T>(list: List<T>): $_list_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly with placeholder 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <T>(list: List<T>): $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly with selectable 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (str: string) => $_string_1;
    <T>(list: List<T>): $_list_1<T>;
    <$SEL extends \\"1\\", $KIND extends \\"list\\">(): <T>(list: List<T>) => $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`generics + multi-parameters should transform correctly without any options 1`] = `
"type $_0 = {
    (str: string): $_string_1;
    <T>(list: List<T>): $_list_1<T>;
};
type $_string_1 = string;
type $_list_1<T> = T | undefined;
"
`;

exports[`merge-non-conflicts should return correctly with mergable types 1`] = `
"type $_00 = {
    <T, U extends List<T>>(fn: (x: T) => T, container: U): $_list_11<T, U>;
    <T, U extends Dictionary<T>>(fn: (x: T) => T, container: U): $_object_11<T, U>;
    <T>(fn: (x: T) => T): $_10<T>;
};
type $_10<T> = {
    <U extends List<T>>(container: U): $_list_11<T, U>;
    <U extends Dictionary<T>>(container: U): $_object_11<T, U>;
};
type $_list_01<T, U extends List<T>> = {
    (fn: (x: T) => T): $_list_11<T, U>;
};
type $_object_01<T, U extends Dictionary<T>> = {
    (fn: (x: T) => T): $_object_11<T, U>;
};
type $_list_11<T, U extends List<T>> = U;
type $_object_11<T, U extends Dictionary<T>> = U;
"
`;

exports[`merge-non-conflicts should return correctly with un-mergable types 1`] = `
"type $_00 = {
    <T, U extends List<T>>(fn: (x: T) => T, container: U): $_list_11<T, U>;
    <T>(fn: (x: T) => T): $_list_10<T>;
    <X, Y extends Dictionary<X>>(fn: (x: X) => X, container: Y): $_object_11<X, Y>;
    <X>(fn: (x: X) => X): $_object_10<X>;
};
type $_list_10<T> = {
    <U extends List<T>>(container: U): $_list_11<T, U>;
};
type $_object_10<X> = {
    <Y extends Dictionary<X>>(container: Y): $_object_11<X, Y>;
};
type $_list_01<T, U extends List<T>> = {
    (fn: (x: T) => T): $_list_11<T, U>;
};
type $_object_01<X, Y extends Dictionary<X>> = {
    (fn: (x: X) => X): $_object_11<X, Y>;
};
type $_list_11<T, U extends List<T>> = U;
type $_object_11<X, Y extends Dictionary<X>> = Y;
"
`;

exports[`non-generics should transform correctly with every option 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    <$SEL extends \\"11\\", $KIND extends \\"number\\">(): (a: number, b: number) => $_number_11;
    (a: number): $_number_10;
    <$SEL extends \\"1\\", $KIND extends \\"number\\">(): (a: number) => $_number_10;
    (_a: PH, b: number): $_number_01;
    <$SEL extends \\"01\\", $KIND extends \\"number\\">(): (_a: PH, b: number) => $_number_01;
    (a: string, b: string): $_string_11;
    <$SEL extends \\"11\\", $KIND extends \\"string\\">(): (a: string, b: string) => $_string_11;
    (a: string): $_string_10;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (a: string) => $_string_10;
    (_a: PH, b: string): $_string_01;
    <$SEL extends \\"01\\", $KIND extends \\"string\\">(): (_a: PH, b: string) => $_string_01;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly with placeholder 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    (a: number): $_number_10;
    (_a: PH, b: number): $_number_01;
    (a: string, b: string): $_string_11;
    (a: string): $_string_10;
    (_a: PH, b: string): $_string_01;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly with selectable 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    <$SEL extends \\"11\\", $KIND extends \\"number\\">(): (a: number, b: number) => $_number_11;
    (a: number): $_number_10;
    <$SEL extends \\"1\\", $KIND extends \\"number\\">(): (a: number) => $_number_10;
    (a: string, b: string): $_string_11;
    <$SEL extends \\"11\\", $KIND extends \\"string\\">(): (a: string, b: string) => $_string_11;
    (a: string): $_string_10;
    <$SEL extends \\"1\\", $KIND extends \\"string\\">(): (a: string) => $_string_10;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;

exports[`non-generics should transform correctly without any options 1`] = `
"type $_00 = {
    (a: number, b: number): $_number_11;
    (a: number): $_number_10;
    (a: string, b: string): $_string_11;
    (a: string): $_string_10;
};
type $_number_10 = {
    (b: number): $_number_11;
};
type $_string_10 = {
    (b: string): $_string_11;
};
type $_number_01 = {
    (a: number): $_number_11;
};
type $_string_01 = {
    (a: string): $_string_11;
};
type $_number_11 = boolean;
type $_string_11 = boolean;
"
`;
